tracer/encoder.go

```diff
@@ -3,63 +3,138 @@ package tracer
 import (
 	"bytes"
 	"encoding/json"
+
+	"github.com/ugorji/go/codec"
 )
 
-// jsonEncoder encodes a list of spans in JSON format.
+// Encoder is a generic interface that expects an Encode() method
+// for the encoding process, and a Read() method that will be used
+// by the http handler
+type Encoder interface {
+	Encode(traces [][]*Span) error
+	Read(p []byte) (int, error)
+	ContentType() string
+}
+
+var mh codec.MsgpackHandle
+
+// msgpackEncoder encodes a list of traces in Msgpack format
+type msgpackEncoder struct {
+	buffer      *bytes.Buffer
+	encoder     *codec.Encoder
+	contentType string
+}
+
+func newMsgpackEncoder() *msgpackEncoder {
+	buffer := &bytes.Buffer{}
+	encoder := codec.NewEncoder(buffer, &mh)
+
+	return &msgpackEncoder{
+		buffer:      buffer,
+		encoder:     encoder,
+		contentType: "application/msgpack",
+	}
+}
+
+// Encode serializes the given traces list into the internal
+// buffer, returning the error if any
+func (e *msgpackEncoder) Encode(traces [][]*Span) error {
+	e.buffer.Reset()
```

Comment:
I believe this is not directly to your msgpack update as the JSON encoder was implemented that way, but it feels weird that buffer is `reset` when calling `Encode()` but not while calling `Decode()`. We could be tricked by this later (forgetting to `reset` before reading for example). I think there's no best way to do it, either call `reset` and mention it in the docs, saves lines in calling code but loses the streaming ability. Or, let the caller be responsible for calling `reset` when needed. In any case, should be mentionned in the docs I think.